target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh

```diff
@@ -20,8 +20,7 @@ platform_do_upgrade_mikrotik_nand() {
 	board_dir=${board_dir%/}
 	[ -n "$board_dir" ] || return
 
-	mtd erase kernel
-	tar xf "$1" ${board_dir}/kernel -O | nandwrite -o "$fw_mtd" -
+	tar xf "$1" ${board_dir}/kernel -O | yafut -d "$fw_mtd" -w -i - -o kernel -m 0755 || return
```

Comment:
Could this fail?
What happens when the Yaffs is partly broken?

Reply:
> Could this fail?

It can, if Yaffs code reaches the conclusion that it is unable to write the `kernel` file to the MTD, for whatever reason.  In particular, this can happen if the MTD partition holding the kernel is using different Yaffs parameters than the defaults assumed by Yafut.  This is why earlier revisions of this PR (e.g. dacf09232647d2742fedc2f59f96d93e7e7732c2) added a separate helper function to `target/linux/ath79/mikrotik/base-files/lib/upgrade/platform.sh` that devices could be switched to once someone verifies that the new kernel update method works fine for a given device.  The feedback I [heard](#discussion_r1149585644) was that MikroTik NAND images are currently broken anyway, so it's okay to switch them to a new tool that may help solve the problem and see what happens.

If the Yaffs parameters used by Yafut are the same as the parameters used by the existing file system on the MTD partition holding the kernel, writing should not fail (assuming of course that the kernel file fits on the given MTD partition).

> What happens when the Yaffs is partly broken?

As it is often the case, it depends.  The most reasonable answer I can give is that what happens in such a case would be similar to mounting the relevant MTD partition using a kernel that has Yaffs support compiled in and running `mv /path/to/kernel /mnt/yaffs-partition/kernel`.  If the file system is "not that broken", Yaffs code may be able to fix it.  Yafut does not contain any extra Yaffs-related logic that the upstream Yaffs code would not already contain.  If things are utterly broken, `mtd erase kernel` and creating a fresh file system may be a feasible last-resort option.

> Is yafut handling a partly damaged file system?

I think my answer to the previous question also applies here.

In short: once the proper set of Yaffs parameters that the bootloader on a given device expects is determined and Yafut is told to use them, I would expect things to be stable indefinitely.

If I misunderstood your questions, please let me know and I'll be happy to explain further.

Reply:
>  What happens when the Yaffs is partly broken?

If the YAFFS is modified by `yafut` on NAND, but the write is incomplete (possible example, kernel too big: not sure if we test this in build, or should test before sysupgrade), device will not boot OpenWrt. Recoverable through BOOTP with button access. If the kernel ELF header is broken, device will normally attempt BOOTP automatically.

---

Neither sure it would work, nor how useful it would be, but we _could_ check if the kernel reads from NAND in `platform_check_image`. Example usage here: https://github.com/openwrt/openwrt/blob/787ecdf66de6798540fb8b756404566341b4733c/target/linux/ipq40xx/base-files/lib/upgrade/platform.sh#L7
`yafut` read should work unless kernel on NAND is broken (example, NAND erased), and then device should not be able to boot NAND anyway, unless `yafut` is having an error reading (example, wrong geometry). From my quick look through https://github.com/openwrt/openwrt/blob/787ecdf66de6798540fb8b756404566341b4733c/package/base-files/files/usr/libexec/validate_firmware_image#L59 it seems sysupgrade would still be forceable, so we could then notify user.

It would make the upgrade take longer for the read, and we are not expecting ath79 NAND devices to vary NAND geometry. I think wait for more feedback before implementing this?

Cheers

Reply:
> Neither sure it would work, nor how useful it would be, but we _could_ check if the kernel reads from NAND in `platform_check_image`. Example usage here:

I think that's quite overkill. Remember that up to that point, we were blindly writing a raw blob (containing a yaffs fs) to the underlying device without any kind of check. Surely a tool that actually handles the filesystem is a major step up already.

> `yafut` read should work unless kernel on NAND is broken (example, NAND erased), and then device should not be able to boot NAND anyway, unless `yafut` is having an error reading (example, wrong geometry). 

Remember that all currently "supported" ath79 NAND devices (128+MB NAND) use the same geometry, as evidenced by the same set of kernel2minor parameters being used for all these devices (and further evidenced by having a single image fitting all devices in the ar71xx era).

My 2c: let's not make this more complicated than it has to be ;)

Reply:
> Neither sure it would work, nor how useful it would be, but we _could_ check if the kernel reads from NAND in `platform_check_image`.

@john-tho Obviously this is not the worst idea - we even briefly discussed it [above](#pullrequestreview-1349537607) - but I am leaning towards the conclusion that this would not be a good safety measure.  In the experiments I run during Yafut development, it happened sometimes (when the Yaffs parameters were messed up or when the code had bugs that were ironed out later on) that the tool seemed to have read a file from the MTD correctly - exit code 0, no obvious errors reported, correct file size - but the *contents* of the file were malformed.  Given that there is no way to know what the *expected* contents of the `kernel` file are, adding more complexity for reading it before attempting to overwrite it does not seem to be worth the benefits.

> Remember that all currently "supported" ath79 NAND devices (128+MB NAND) use the same geometry, as evidenced by the same set of kernel2minor parameters being used for all these devices (and further evidenced by having a single image fitting all devices in the ar71xx era).

@f00b4r0 This is a good point.  The only thing I can think of that could go wrong is that the MTD partition holding the kernel could theoretically use a different ECC layout than the one selected by the Linux kernel for the NAND chip.  Given that I only own two (very similar) devices, it is impossible for me to assess whether that is something that happens out there.

Reply:
> @f00b4r0 This is a good point. The only thing I can think of that could go wrong is that the MTD partition holding the kernel could theoretically use a different ECC layout than the one selected by the Linux kernel for the NAND chip. 

I don't see how: this would result in failed mtd writes regardless of underlying filesystem. The fact that a single image was suitable to work with all ar71xx NAND devices suggests there is no such problem. Again, I think we shouldn't speculate too much, or this PR will never get merged.

In my opinion this PR is **ready for merge now**: the resulting system is *better in every way* to the prior alternative (images not built and potential brick due to lack of wear-management).

Reply:
I could continue the ECC discussion, but supporting your conclusion presented above feels more productive :-)  Perfect is indeed the enemy of good.